### PR TITLE
[sharktank] Add tests for pipeline-parallel Llama 405B and Toy f4 vs non-parallel

### PR DESCRIPTION
**Pull request by sogartar**
Add tests to compare perplexity of PP vs non-PP with a small set of random tokens.

The Llama 405B f4 test is for a pruned (removed layers) model as we don't care if the perplexity is high. Just that it is the same against the reference.

### CODE REVIEW

**Comment by sogartar** [comment]
I plan to migrate this to the preshuffled model variant at a subsequent PR.

File: sharktank/conftest.py [hunk 1]
```diff
@@ -329,6 +343,14 @@ def model_artifacts(request: FixtureRequest) -> dict[str, str]:
...
         request, "--llama3-70b-f8-model-path", "llama3_70b_f8_model"
     )
+    model_path["llama3_1_405b_instruct_f4_model_path"] = set_fixture_from_cli_option(
+        request,
+        "--llama3-1-405b-instruct-f4-model-path",
+        "llama3_1_405b_instruct_f4_model_path",
+    )
+    model_path["llama3_1_405b_tokenizer_path"] = set_fixture_from_cli_option(
+        request, "--llama3-1-405b-tokenizer-path", "llama3_1_405b_tokenizer_path"
+    )
```
Comment (Alex-Vasile) [inline sharktank/conftest.py:353]: This is quickly getting out of hand. We need to change this to not continue with the combinatorial explosion.

File: .github/workflows/ci-sharktank-nightly.yml [hunk 2]
```diff
@@ -21,9 +21,9 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
-  flux-test:
+  nightly-mi300x:
```
Comment (Alex-Vasile) [inline .github/workflows/ci-sharktank-nightly.yml:24]: `nightly-mi325x`?

File: sharktank/tests/models/llama/test_llama.py [hunk 3]
```diff
@@ -61,6 +71,57 @@ def testUnsharded(self):
...
+    device = torch.device("cuda")
+    batch_size = 4
+    prune_to_block_count = 3
+    pipeline_parallelism_size = 2
+
+    parameters_path = model_artifacts["llama3_1_405b_instruct_f4_model_path"]
+    dataset = Dataset.load(parameters_path)
+
+    reference_config = LlamaModelConfig.from_dataset(dataset)
+    # TODO: remove when the IRPA has the correct value
```
Comment (Alex-Vasile) [inline sharktank/tests/models/llama/test_llama.py:95]: Please move to the pre-shuffled IRPA for this PR.

This PR is still blocked by: https://github.com/iree-org/iree-turbine/pull/1157.

File: sharktank/sharktank/models/llm/testing.py [hunk 4]
```diff
@@ -1,93 +1,164 @@
...
-        size=[batch_size, batch_seq_len],
-        dtype=torch.int32,
-        device=model.device,
+
+    pp_config = deepcopy(reference_config)
+    pp_config.parallelism_config = ParallelismConfig.default_config(
+        block_count=reference_config.hp.block_count,
+        pp=pipeline_parallelism_size,
     )
+    pp_theta = Theta(reference_theta.flatten())
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:134]: Is this not going to create 2 copies of the weights in memory?

File: sharktank/sharktank/models/llm/testing.py [hunk 5]
```diff
@@ -1,93 +1,164 @@
...
-            2 * model.config.block_seq_stride,
-            model.config.hp.context_length,
+class LlmPerplexityCompare:
+    """Compare the perplexity of one implementation against another.
+    Checks that the 2 models' results do not deviate for each other."""
+
+    def __init__(
+        self,
+        make_target_model: Callable[[], TorchInstance],
+        make_reference_model: Callable[[], TorchInstance],
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:60]: I don't understand these variable names. Why not `target_model` and `reference_model`?

File: sharktank/sharktank/models/llm/testing.py [hunk 6]
```diff
@@ -1,93 +1,164 @@
...
-        high=min(
-            2 * model.config.block_seq_stride,
-            model.config.hp.context_length,
+class LlmPerplexityCompare:
+    """Compare the perplexity of one implementation against another.
+    Checks that the 2 models' results do not deviate for each other."""
+
+    def __init__(
+        self,
+        make_target_model: Callable[[], TorchInstance],
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:59]: Is this only going to work with TorchInstances?

File: sharktank/sharktank/models/llm/testing.py [hunk 7]
```diff
@@ -1,93 +1,164 @@
...
+        return prefill_results, decode_results
+
+    def _assert_close(
+        self,
+        actual: list[LlmPerplexityEval.Result],
+        expected: list[LlmPerplexityEval.Result],
+    ):
+        actual_scores = torch.tensor([r.score for r in actual], dtype=torch.float32)
+        expected_scores = torch.tensor([r.score for r in expected], dtype=torch.float32)
+        assert_tensor_close(actual_scores, expected_scores, atol=1e-2, rtol=1e-2)
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:104]: These tolerances are quiet large and hardcoded.

File: sharktank/sharktank/types/tensors.py [hunk 8]
```diff
@@ -869,6 +869,14 @@ def __init__(
         super().__init__(name=name, shape=shape, layout_type=type(layout))
         self.layout = layout
 
+    def to(self, *args, **kwargs) -> "PlanarQuantizedTensor":
```
Comment (Alex-Vasile) [inline sharktank/sharktank/types/tensors.py:872]: This doesn't belong in this PR, should be seperate.

File: sharktank/conftest.py (hunk 1 above)
Comment (sogartar) [reply sharktank/conftest.py:353]: Maybe we can have a single CLI option like `--sharktank-test-data-path` for the test files and expect a standard directory structure inside across all machines.

File: sharktank/sharktank/models/llm/testing.py (hunk 6 above)
Comment (sogartar) [reply sharktank/sharktank/models/llm/testing.py:59]: Actually it should work on `IreeInstance`. I will change it.

File: sharktank/conftest.py [hunk 9]
```diff
@@ -185,6 +185,20 @@ def pytest_addoption(parser):
         default=None,
         help="Llama3.1 70b f8 model path",
     )
+    parser.addoption(
+        "--llama3-1-405b-tokenizer-path",
+        type=Path,
+        action="store",
+        default=None,
+        help="Llama3.1 405b tokenizer path. Instruct and Vanilla have the same tokenizer.",
```
Comment (archana-ramalingam) [inline sharktank/conftest.py:193]: We need to confirm this, as I remember 405b fp4 and fp16 tokenizers were different causing numeric issues. (both being instruct versions)

File: sharktank/sharktank/models/llm/testing.py [hunk 10]
```diff
@@ -1,93 +1,164 @@
...
+        )
+
+        instance = LlmInstance(
+            model_instance=model,
+            page_sizes=page_sizes,
+            block_seq_stride=model.config.block_seq_stride,
+            block_count=page_count,
+            decode_topk_logits=None,
+        )
+        perplexity_eval = instance.make_perplexity_eval()
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/llm/testing.py:90]: I see an overlap with existing ppl [testing](https://github.com/nod-ai/shark-ai/blob/main/sharktank/sharktank/tools/eval_llm_model.py#L36), and can be consolidated?

File: sharktank/sharktank/types/tensors.py (hunk 8 above)
Comment (sogartar) [reply sharktank/sharktank/types/tensors.py]: Moved it to a new PR https://github.com/nod-ai/shark-ai/pull/2376.

File: sharktank/sharktank/models/llm/testing.py (hunk 10 above)
Comment (sogartar) [reply sharktank/sharktank/models/llm/testing.py:87]: I reworked the class a bit to accept `LlmInstance`.
This class is a bit more bare bones. It does not use a dataset and tokenizer and assumes everything is in 1 batch.
It is more about the side-by-side comparison of 2 LLM models.

File: sharktank/conftest.py (hunk 9 above)
Comment (sogartar) [reply sharktank/conftest.py]: I reworked this part to have a single test data dir CLI flag.

File: sharktank/integration/models/llama/llama_integration_test.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        test_data_dir
+        / "ossci-models/llama_3_1/405b/fp4/fp4_preshuffled_2025_09_12.irpa"
+    )
+    dataset = Dataset.load(parameters_path)
+
+    reference_config = LlamaModelConfig.from_dataset(dataset)
+
+    assert (
+        reference_config.fake_quant == True
+    ), "TODO: remove fake_quant fix below when it has the correct value"
```
Comment (Alex-Vasile) [inline sharktank/integration/models/llama/llama_integration_test.py:58]: Do you have an issue for this?

File: .github/workflows/ci-sharktank-nightly.yml [hunk 12]
```diff
@@ -80,6 +81,22 @@ jobs:
...
 
+      - name: Run integration tests
+        run: |
+          source ${VENV_DIR}/bin/activate
+          pytest \
+            --test-data-dir=/shark-dev \
+            --iree-hal-target-device=hip \
+            --iree-hip-target=gfx942 \
+            --iree-device=hip://0 \
+            --device='cuda:0' \
```
Comment (Alex-Vasile) [inline .github/workflows/ci-sharktank-nightly.yml:92]: Why are these being specified if only the tests are only ever run on CPU?

File: sharktank/sharktank/models/llm/testing.py [hunk 13]
```diff
@@ -1,93 +1,185 @@
...
+            actual_scores, expected_scores, rtol=self.rtol, atol=self.atol
+        )
+
+
+def run_perplexity_test_pipeline_parallel_eager_vs_eager(
+    reference_theta: Theta,
+    reference_config: LlamaModelConfig,
+    tokens: list[list[int]],
+    atol: float = 0,
+    rtol: float = 0,
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:111]: I know I originally asked this to be paramaterized, but I was wrong.

These should be hardcoded as 0 when calling `LlmPerplexityCompare()` below.

File: sharktank/sharktank/models/llm/testing.py [hunk 14]
```diff
@@ -1,93 +1,185 @@
...
+def clip_llm_block_count(
+    theta: Theta, config: LlamaModelConfig, block_count: int
+) -> tuple[Theta, LlamaModelConfig]:
+    """Remove all trailing layers/blocks from the theta to align the desired block/layer count."""
+    assert (
+        config.pipeline_parallelism_size == 1 and config.tensor_parallelism_size == 1
+    ), "Not supported"
+
+    config = deepcopy(config)
+    config.hp.block_count
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:177]: This line doesn't seem to do anything.

File: sharktank/sharktank/ops/default_impls.py [hunk 15]
```diff
@@ -575,12 +575,11 @@ def index_copy__default(
...
-        not torch.compiler.is_compiling()
-        and inout_as_torch.is_cpu
-        and inout_as_torch.dtype in [torch.float8_e4m3fnuz, torch.float8_e4m3fn]
-    ):
-        # PyTorch does not have eager implementation for float8_e4m3fnuz in CPU.
+    if not torch.compiler.is_compiling() and inout_as_torch.dtype in [
+        torch.float8_e4m3fnuz,
+        torch.float8_e4m3fn,
+    ]:
+        # PyTorch/PyTorch ROCm does not have eager implementation for various dtypes for CPU/GPU.
```
Comment (Alex-Vasile) [inline sharktank/sharktank/ops/default_impls.py:582]: When is this failing?

Also this should ideally be in a separate PR.

**Review by Alex-Vasile — CHANGES_REQUESTED** [review]

File: sharktank/sharktank/ops/default_impls.py (hunk 15 above)
Comment (sogartar) [reply sharktank/sharktank/ops/default_impls.py:582]: If you run in eager on GPU. May not have `index_copy_` for these dtypes. I encountered it.

Comment (sogartar) [reply sharktank/sharktank/ops/default_impls.py:582]: Here is a separate PR https://github.com/nod-ai/shark-ai/pull/2389.

File: sharktank/sharktank/models/llm/testing.py (hunk 13 above)
Comment (sogartar) [reply sharktank/sharktank/models/llm/testing.py]: I hardcoded them in `run_perplexity_test_pipeline_parallel_eager_vs_eager`.

File: sharktank/sharktank/models/llm/testing.py [hunk 16]
```diff
@@ -1,93 +1,182 @@
...
+    theta: Theta, config: LlamaModelConfig, block_count: int
+) -> tuple[Theta, LlamaModelConfig]:
+    """Remove all trailing layers/blocks from the theta to align the desired block/layer count."""
+    assert (
+        config.pipeline_parallelism_size == 1 and config.tensor_parallelism_size == 1
+    ), "Not supported"
+
+    config = deepcopy(config)
+    # Make sure block_count derivative values are recomputed.
+    config.parallelism_config = None
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:176]: Why remove the parallelism config?

File: sharktank/sharktank/models/llm/testing.py [hunk 17]
```diff
@@ -1,93 +1,185 @@
...
+    """Remove all trailing layers/blocks from the theta to align the desired block/layer count."""
+    assert (
+        config.pipeline_parallelism_size == 1 and config.tensor_parallelism_size == 1
+    ), "Not supported"
+
+    config = deepcopy(config)
+    config.hp.block_count
+    # Make sure block_count derivative values are recomputed.
+    config.parallelism_config = None
+    config.__post_init__()
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/models/llm/testing.py]: I made it like you suggested. I also made this fix https://github.com/nod-ai/shark-ai/pull/2393.

File: sharktank/sharktank/models/llm/testing.py (hunk 14 above)
Comment (sogartar) [reply sharktank/sharktank/models/llm/testing.py]: This actually a real error that needed to be
```
config.hp.block_count = block_count
```

File: sharktank/sharktank/models/llm/testing.py (hunk 16 above)
Comment (sogartar) [reply sharktank/sharktank/models/llm/testing.py]: It is because the number of blocks gets reduced and this causes shape discrepancies downstream. I think I patched this config molestation. Its ugly.
Maybe if I get https://github.com/nod-ai/shark-ai/pull/2393 landed it will be a bit better.

File: sharktank/integration/models/llama/llama_integration_test.py (hunk 11 above)
Comment (sogartar) [reply sharktank/integration/models/llama/llama_integration_test.py]: I added it.

File: sharktank/sharktank/models/llm/testing.py [hunk 18]
```diff
@@ -1,93 +1,187 @@
...
+    assert (
+        config.pipeline_parallelism_size == 1 and config.tensor_parallelism_size == 1
+    ), "Not supported"
+
+    config = deepcopy(config)
+    config.hp.block_count = block_count
+    # Make sure block_count derivative values are recomputed.
+    config_as_props = config.to_properties()
+    del config_as_props["parallelism_config"]
+    del config_as_props["tensor_parallelism_size"]
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/testing.py:179]: Why is this deletion needed?

Comment (sogartar) [reply sharktank/sharktank/models/llm/testing.py:179]: I guess specifically
```
del config_as_props["tensor_parallelism_size"]
```

is defensive programming. I will remove this one.

**Comment by sogartar** [comment]
Something is wrong with `write_timestep` in the KV cache.
